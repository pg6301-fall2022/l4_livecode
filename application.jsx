import * as React from "react";
import * as ReactDOM from "react-dom";

export function ListMovies(){
    return <div>
        <h1>Something else basic, but in a different file</h1>
    </div>
}