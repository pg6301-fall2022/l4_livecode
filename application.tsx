import * as React from "react";
import * as ReactDOM from "react-dom";

import {useState} from "react";

export function ListMovies(){
    return <div>
        <h1>Something else basic, but in a different file</h1>
    </div>
}

interface Movie {
    title: String;
    year: String;
    plot: String;
}

export function NewMovieForm(
    {onAddMovie,} : {onAddMovie(mov: Movie): void }
){
    const [title, setTitle] = useState("");
    const [year, setYear] = useState("");
    const [plot, setPlot] = useState("");

    function handleSubmit() {
        onAddMovie({title, year, plot});
    }

    return <form onSubmit={handleSubmit}>
        <h1> New Movie </h1>
        <div>
            Title: <input data-testid="title" value={title} onChange={e => setTitle(e.target.value)}/>
        </div>
        <div>
            Year: <input data-testid="year" value={year} onChange={e => setYear(e.target.value)}/>
        </div>
        <div>
            Plot: <textarea data-testid="plot" value={plot} onChange={e => setPlot(e.target.value)}/>
        </div>
    </form>
}