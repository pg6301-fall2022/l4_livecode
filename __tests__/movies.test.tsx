import * as React from "react";
import * as ReactDOM from "react-dom";

import {ListMovies, NewMovieForm} from "../application";
import {act, Simulate} from "react-dom/test-utils";
import {createRoot} from "react-dom/client";
import {on} from "process";

describe("movie app tests", () => {

    it("shows movie list", ()=> {
        const element = document.createElement('div');

        const root = createRoot(element);
        act( () => root.render(
            <ListMovies />
        ));

        //ReactDOM.render(<ListMovies/>, element);

        expect(element.querySelector("h1")?.innerHTML).toEqual("Something else basic, but in a different file");
        expect(element.innerHTML).toMatchSnapshot();
    });

    it("new movie form shows", () => {
        const element = document.createElement("div");
        const onAddMovieMocked = jest.fn(); // Mock function

        const root = createRoot(element);
        act(() => {
           root.render(<NewMovieForm onAddMovie={onAddMovieMocked} />);
        });

        //ReactDOM.render(<NewMovieForm onAddMovie={onAddMovieMocked}/>, element);

        expect(element.innerHTML).toMatchSnapshot();
    });

    it("form test", () => {
        const onAddMovieMocked = jest.fn(); // Mock function

        // on submit this will be called instead of the actual function (if that even exists).

        const element = document.createElement("div");
        const root = createRoot(element);

        act( () => {
           root.render(<NewMovieForm onAddMovie={onAddMovieMocked} />);
        });

        //ReactDOM.render(<NewMovieForm onAddMovie={onAddMovieMocked}/>, element);

        act( () =>
            Simulate.change(
            element.querySelector("[data-testid=title]")!,
            {target: {value: "Movie1"}} as any
        )
        );
        // I simulate a change in the input on the form

        act( () =>
        Simulate.change(
            element.querySelector("[data-testid=year]")!,
            {target: {value: "2022"}} as any
        )
        );

        // At this point, my form has some simulated data.

        act( () =>
            Simulate.submit(element.querySelector("form")!) // Submit the simulated data
        );

        //expect(onAddMovieMocked).toHaveBeenCalled();

        expect(onAddMovieMocked).toHaveBeenCalledWith({
            title: "Movie1",
            year: "2022",
            plot: ""
            });

    });

});