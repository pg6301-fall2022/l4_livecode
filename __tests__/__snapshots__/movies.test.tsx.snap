// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`movie app tests new movie form shows 1`] = `"<form><h1> New Movie </h1><div>Title: <input data-testid="title" value=""></div><div>Year: <input data-testid="year" value=""></div><div>Plot: <textarea data-testid="plot"></textarea></div></form>"`;

exports[`movie app tests shows movie list 1`] = `"<div><h1>Something else basic, but in a different file</h1></div>"`;
